[Catalyst] Fix OnPlatform and OnIdiom on macOS

jsuarezruiz opened the pull request:
### Description of Change ###

Fix OnPlatform on macOS.

- fixes #3758 

### PR Checklist ###

- [x] Targets the correct branch 
- [ ] Tests are passing (or failures are unrelated)

#### Does this PR touch anything that might affect accessibility?
- No

jsuarezruiz committed:
Fix RuntimePlatform in macOS

Redth commented:
Hmm I actually wonder if we should add a new one for MacCatalyst... In the event that we ever do a macOS backend based on AppKit/Cocoa, would we want this to report differently?

hartez commented:
Or should this be `OnPlatform == iOS` and `OnIdiom == Desktop`?

Redth commented:
> Or should this be `OnPlatform == iOS` and `OnIdiom == Desktop`? And then AppKit/Cocoa would be `OnPlatform == AppKit` (or 'Cocoa') and `OnIdiom == Desktop`?

Yeah I'm inclined to pick this option.  I think the fact that this is pretty much iOS (or rather iPadOS if we really want to be correct), but the factor (Idiom) is Desktop.  I think it would be fine to even choose `MacOS` as the Platform with `iPadOS` or `iOS` as the Platform ...

vanka78bg commented:
We are tracking this issue as well, as the bug is impacting our Mac OS support. I have a few questions though. Changing `Device.RuntimePlatform` to return `macOS` instead of `iOS` would break existing code, e.g.:

```xml
<Label TextColor="{OnPlatform Android='Blue', iOS='Red'}" />
```

After the change, the above styling would not be applied on Mac OS. To correct this, we would have to modify the XAML like this:

```xml
<Label TextColor="{OnPlatform Android='Blue', iOS='Red', macOS='Red'}" />
```

This is not such a big issue, as MacCatalyst support is relatively new and the existing UI would most likely require some tweaking anyway. The bigger problem would arise if we keep `Device.RuntimePlatform` returning `iOS` on Mac OS, but change `Device.Idiom` to return `Desktop` instead. For instance, if we want to set a specific styling for Mac OS like this:

```xml
<Label TextColor="{OnIdiom Desktop='Red'}" />
```

The above styling would be applied to both Windows and Mac OS, which is not desirable. What we need is to apply the styling when `OnPlatform` is `iOS` ***AND*** `OnIdiom` id `Desktop`. This is not currently possible to express in XAML.

For the above reason my vote goes to change `Device.RuntimePlatform` to return `MacCatalyst` and `Device.Idiom` to return `Desktop`. Then we can add another option to the `OnPlatform` markup extension for `MacCatalyst`. In my opinion, that would be the safest option, provided you have plans for a true Mac OS support in the future.

Redth commented:
> We are tracking this issue as well, as the bug is impacting our Mac OS support. I have a few concerns though. Changing `Device.RuntimePlatform` to return `macOS` instead of `iOS` would break existing code, e.g.:
> 
> ```
> <Label TextColor="{OnPlatform Android='Blue', iOS='Red'}" />
> ```
> 
> After the change, the above styling would not be applied on Mac OS. To correct that, we would have to modify the XAML like this:
> 
> ```
> <Label TextColor="{OnPlatform Android='Blue', iOS='Red', macOS='Red'}" />
> ```
> 
> This is not such a big issue, as MacCatalyst support is relatively new and the existing UI would most likely require some tweaking anyway. The bigger problem would arise if we keep `Device.RuntimePlatform` returning `iOS` on Mac OS, but change `Device.Idiom` to return `Desktop` instead. For instance, if we want to set a specific styling for Mac OS like this:
> 
> ```
> <Label TextColor="{OnIdiom Desktop='Red'}" />
> ```
> 
> The above styling would be applied to both Windows and Mac OS, which is not desirable. What we need is to apply the styling when `OnPlatform` is `iOS` **AND** `OnIdiom` is `Desktop`. This is not currently possible to express in XAML, however.
> 
> For the above reasons, my vote goes to change `Device.RuntimePlatform` to return `MacCatalyst` and `Device.Idiom` to return `Desktop`. Then you can add another option to the `OnPlatform` markup extension for `MacCatalyst`. In my opinion, that would be the safest option, provided you have plans for a true Mac OS support in the future.

These are good points and I agree with your analysis.

I took a look at what Apple does with swift... and in 5.5 they have:
`os()` which can be any of: `macOS`, `iOS`, `watchOS`, `tvOS`, `Linux`, or `Windows`.
`targetEnvironment()` which can be `simulator`, `macCatalyst`

OS seems reasonable, as you're still in fact on macOS whether you're in an AppKit or MacCatalyst app.  
TargetEnvironment is more peculiar as the values don't seem particularly helpful outside of iOS/MacCatalyst scenarios.

Now, .NET 6 gives us the following in `System.OperatingSystem`:

```
public static bool IsOSPlatform (string platform)
public static bool IsWindows ()
public static bool IsWatchOS ()
public static bool IsTvOS ()
public static bool IsMacCatalyst ()
public static bool IsMacOS ()
public static bool IsIOS ()
public static bool IsAndroid ()
public static bool IsFreeBSD ()
public static bool IsLinux ()
public static bool IsBrowser ()
```

What `RuntimePlatform` I think really should equate to the `System.OperatingSystem` methods.
If you want to generically know if you're on `macOS` (either via `net6.0-macos` or `net6.0-maccatalyst`), you can simply check both conditions.  In terms of XAML you would basically duplicate your markup to have the same value for both `OnPlatform` values, though in practice I suspect it's unlikely to use the same values for MacCatalyst/macOS in such shared code in the future.

To the point about iOS not working on MacCatalyst in cases like `OnPlatform`, as you mentioned @vanka78bg this is a new platform so I think it's reasonable, and I wouldn't expect a new platform to fall back to a different platform's values typically anyway.

My vote is `RuntimePlatform => "MacCatalyst"`

Redth committed:
Add MacCatalyst to Device

Redth committed:
Fix RuntimePlatform for Apple platforms

Redth committed:
Add tvOS to Device

jsuarezruiz commented:
/azp run

Redth committed:
Merge branch 'main' into fix-3758

Redth committed:
Also change Essentials' DevicePlatform to match

Redth committed:
Check Essentials DeviceInfo for Platform too

Eventually Device.RuntimePlatform should be removed

mattleibow commented on src/Essentials/src/DeviceInfo/DeviceInfo.ios.tvos.watchos.cs:
```diff
@@ -33,8 +33,10 @@ static string GetModel()
-#if __IOS__
+#if __IOS__ || IOS
 			DevicePlatform.iOS;
+#elif __MACCATALYST__ || MACCATALYST
```
We may have to swap this because maccat is also ios, so we may have to put maccat first.

mattleibow committed:
Merge branch 'main' into fix-3758

Redth committed:
Update DeviceInfo.ios.tvos.watchos.cs

Redth commented on src/Essentials/src/DeviceInfo/DeviceInfo.ios.tvos.watchos.cs:
```diff
@@ -33,7 +33,9 @@ static string GetModel()
+#if __MACCATALYST__ || MACCATALYST
+			DevicePlatform.MacCatalyst;
+#elif __IOS__ || IOS
 			DevicePlatform.iOS;
```
@mattleibow  like this?

mattleibow committed:
Merge remote-tracking branch 'origin/main' into fix-3758

Redth committed:
Platform value to string

Redth committed:
Merge branch 'main' into fix-3758

mattleibow commented:
Does this also need to be updated in the OnPlatformExtension?

mattleibow commented:
@StephaneDelcroix  probably should give this a look just in case we missed something

mattleibow committed:
Add maccatalyst and remove essentials

I removed essentials as that sometimes messed with the tests because it is not mocked yet

mattleibow commented:
@Redth I removed the check in OnPlatform with Essentials because tests are failing as a result of it not being mocked. We need to  work on getting more essentials APIs to be mockable - both for our tests, but also for user tests.

mattleibow committed:
Merge remote-tracking branch 'origin/main' into fix-3758

mattleibow committed:
Added this

Redth committed:
Set idiom to Desktop if Catalyst

Redth committed:
Merge branch 'fix-3758' of https://github.com/dotnet/maui into fix-3758

mattleibow reviewed (approved):
LGTM

sbeca commented on src/Compatibility/Core/src/iOS/Forms.cs:
```diff
@@ -637,8 +639,12 @@ public async Task<Stream> GetStreamAsync(Uri uri, CancellationToken cancellation
 			}
 
-#if __MOBILE__
+#if MACCATALYST
```
Every other check in this PR uses `#if MACCATALYST || __MACCATALYST__` but this line only uses `#if MACCATALYST`. Is that an issue?

sbeca commented on src/Essentials/src/DeviceInfo/DeviceInfo.ios.tvos.watchos.cs:
```diff
@@ -45,6 +47,8 @@ static DeviceIdiom GetIdiom()
 		{
 #if __WATCHOS__
             return DeviceIdiom.Watch;
+#elif MACCATALYST || __MACCATALYST
```
Should `__MACCATALYST` instead be `__MACCATALYST__`?

Redth replied:
Yes looks like a typo, however it shouldn't matter, the `MACCATALYST` is the proper one for net6, and `__MACCATALYST__` was really for legacy TFM's plus VSMac didn't recognize the `MACCATALYST` one until very recent internal builds.

mattleibow commented:
Totally... Something went wrong here. I am sure at some point it was correct. Thanks for catching this!
